### ISSUE TITLE
redundant setting of password and db in a pool

### ISSUE THREAD

**Issue by drobert** [xetorthio/jedis#1489]
`JedisPool` internally creates a `JedisFactory` with no means of overriding the factory. The factory returns instances of `Jedis` via its internal `makeObject()` method. Related, `Jedis`'s internal `Client` (`redis.clients.jedis.Client`), while protected, has no setter or constructor argument allowing it to be replaced. (I bring up the inability to override values not because it's troublesome, though it is, but because it implies a closed system with behavior controller by the library rather than the end user.)

Within the `JedisFactory`, method `makeObject()` attempts to set a password and set the selected database if they are non-standard values, which seems reasonable. However, the underlying `Client` extends `BinaryClient` which has the following in its `connect()` method:

```java
  @Override
  public void connect() {
    if (!isConnected()) {
      super.connect();
      if (password != null) {
        auth(password);
        getStatusCodeReply();
      }
      if (db > 0) {
        select(Long.valueOf(db).intValue());
        getStatusCodeReply();
      }
    }
  }
```

In short: when a Jedis instance is newly-instantiated within a pool,  the factory calls the underlying client's connect, which sets the database and password. The pool itself then sets the database and password a second time, superfluously. 

### Expected behavior

- Authentication should be established once per initial connection. 
- Database should be set once on initial connection, and reset when the object is returned to the pool

### Actual behavior

- Authentication is executed twice when an object is newly added to the connection pool
- Database is set twice when an object is newly added to the connection pool

### Redis / Jedis Configuration

N/A

#### Jedis version:

2.7.3

#### Redis version:

N/A

#### Java version:

N/A

**Comment by marcosnils** [comment]
@drobert this is incorrect, when you initialize a JedisPool with a password like this:

`JedisPool pool = new JedisPool(new JedisPoolConfig(), "localhost", 6379, 2000,"foobared");`

Then the password will be set at the `pool` level and not at the `Jedis` level, so https://github.com/xetorthio/jedis/blob/master/src/main/java/redis/clients/jedis/BinaryClient.java#L100 will return null and only `JedisFactory` will issue the `auth` command accordingly. 

You can verify this with a simple program and `monitor` the commands sent to the redis server.

**Comment by drobert** [comment]
Indeed, this indeed up being the result of an internally modified Pool where Jedis instances were instantiated by passing JedisShardInfo instances. In the case of `new Jedis(JedisShardInfo)`, any database and password in the shard info are passed along to the client, and this scenario is possible. However you're correct that there is no configuration out of the box that leads to this.
